File: src/webcore/model/abstractmodel.ts
"use strict";

export class AbstractModel {
	public _wasm_model: any = null;

	constructor(context: any) {
		this._wasm_model = new context.module.AbstractModel();
	}

	public getProps(): any {
		return this._wasm_model.getProps();
	}

	public getPropBy(key: string): any {
		return this._wasm_model.getProps()[key];
	}

	public setProp(key, value): void {
		this._wasm_model.setProp(key, value);
	}
};

File: src/webcore/model/abstractmodel.cpp
#include <sway/webcore/model/abstractmodel.h>

NAMESPACE_BEGIN(sway)
NAMESPACE_BEGIN(webcore)
NAMESPACE_BEGIN(model)

ModelSmartPtr_t AbstractModel::create() {
	return std::make_shared<AbstractModel>();
}

AbstractModel::AbstractModel()
	: _data(emscripten::val::object()) {
	// Empty
}

AbstractModel::~AbstractModel() {
	// Empty
}

void AbstractModel::setProp(const std::string & key, const emscripten::val & value) {
	_data.set(key, value);
	notify();
}

emscripten::val AbstractModel::getProps() const {
	return _data;
}

NAMESPACE_END(model)
NAMESPACE_END(webcore)
NAMESPACE_END(sway)
